### PR TITLE
feat: create a minimal docker image and github workflow for building it when pushed to master branch

### PR DESCRIPTION
**Pull request by crebsy**
# What I did

- Created a docker image which contains `rindexer` and the graphql server binary with docker
- Github workflow that builds and tags it on ghcr.io when a branch starting with `release/` is merged/pushed to `master`

### CODE REVIEW

**Comment by joshstevens19** [comment]
good PR @crebsy - can we add some documentation?

**Comment by crebsy** [comment]
@joshstevens19
I've just added a small section for Docker inside the README and also packaged the static binary for the graphql server into the final image.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM --platform=linux/amd64 clux/muslrust:stable as builder
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
+RUN mkdir /app
```
Comment (o-az) [inline Dockerfile:3]:
```suggestion
```

`WORKDIR` instruction automatically creates directory if it doesn't exist

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+WORKDIR /app
+COPY . .
+RUN rustup target add x86_64-unknown-linux-musl
+RUN cargo build --release --target x86_64-unknown-linux-musl
+RUN strip /app/target/x86_64-unknown-linux-musl/release/rindexer_cli
+
+FROM --platform=linux/amd64 node:16 as node-builder
```
Comment (o-az) [inline Dockerfile:10]:
```suggestion
FROM --platform=linux/amd64 node:lts as node-builder
```

Node.js v16 is no longer maintained and doesn't receive security updates
https://endoflife.date/nodejs

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+RUN mkdir /app
```
Comment (o-az) [inline Dockerfile:11]:
```suggestion
```

**Review by o-az — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (crebsy) [reply Dockerfile]: thanks, removed!

File: Dockerfile (hunk 2 above)
Comment (crebsy) [reply Dockerfile]: yeah I've seen that it was used inside the package.json for building the graphql binary but you're right we should bump it

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
+WORKDIR /app
+COPY . .
+RUN rustup target add x86_64-unknown-linux-musl
+RUN cargo build --release --target x86_64-unknown-linux-musl
+RUN strip /app/target/x86_64-unknown-linux-musl/release/rindexer_cli
+
+FROM --platform=linux/amd64 node:lts as node-builder
+WORKDIR /app
+COPY . .
```
Comment (o-az) [inline Dockerfile:11]: I think copying the files/dirs needed specifically will prevent unwanted files from being copied over

alternatively you can ignore everything in `.dockerignore` except for the files you want to copy over. Example

```dockerignorr
# ignore everything
*

# except the following
!scripts/
!Dockerfile
!.dockerignore
!rindexer
```

https://github.com/o-az/rindexer-railway/blob/main/.dockerignore

Comment (crebsy) [reply Dockerfile:11]: true but in the end, we'll copy over only the two static binaries into the final alpine image and just publish that to the registry.

**Comment by o-az** [comment]
I was able to run this locally using just the `Dockerfile`. So I think you can merge this and I will update my pr to include railway template @joshstevens19

**Review by o-az — APPROVED** [review]

**Review by joshstevens19 — APPROVED** [review]
Epic work

**Pull request merged**